<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to ampvis2 • ampvis2</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">ampvis2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ampvis2.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MadsAlbertsen/ampvis2">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Introduction to ampvis2</h1>
            
          </div>

    
    
<div class="contents">
<p>This is a short guide to the basics of ampvis2, how to load data, as well as a few basic visualisation functions using example data. A complete explanation of all functions, their purpose, different arguments and examples can be found in the <a href="../reference/index.html">Reference</a> tab. First, install ampvis2 as described on the front page.</p>
<div id="loading-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading data</h2>
<p>The most notable change with ampvis2 is that phyloseq is no longer used to handle the data, which makes it simpler to understand and work with, and lighter. The data is now simply stored as a list containing 3 dataframes (4 if reference sequences are provided) named <code>"metadata"</code>, <code>"otutable"</code> and <code>"tax"</code> (short for taxonomy). The data ampvis2 needs is an OTU-table generated with fx <a href="https://github.com/MadsAlbertsen/16S-analysis/tree/master/data.generation">workflow scipts v.4+</a> and a metadata sheet containing information about the samples.</p>
<p>More precisely, the OTU-table provided is expected to be a raw CSV-table with OTU names/ID’s in rows and sample ID’s in columns, where numbers in the corresponding cells represent the read counts of the particular OTU in each sample. The last 7 columns of the OTU-table are expected to be the taxonomy of the corresponding OTU’s (Kingdom-&gt;Species). The metadata needs to have the same sample ID’s as in the OTU-table in the first column and then the remaining columns contain information about the samples, for example where the sample was taken, date, pH, temperature etc. First, import your data into R dataframes using an appropriate read function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myotutable &lt;-<span class="st"> </span><span class="kw">read.csv2</span>(<span class="st">"data/otutable.csv"</span>, 
                        <span class="co">#sep = "\t",</span>
                        <span class="co">#dec = ".",</span>
                        <span class="dt">fill =</span> <span class="ot">TRUE</span>,
                        <span class="dt">header =</span> <span class="ot">TRUE</span>,
                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>,
                        <span class="dt">check.names =</span> <span class="ot">FALSE</span>,
                        <span class="dt">row.names =</span> <span class="dv">1</span>) %&gt;%<span class="st"> </span><span class="kw">as.data.frame</span>()
mymetadata &lt;-<span class="st"> </span><span class="kw">read_excel</span>(<span class="st">"data/metadata.xlsx"</span>,
                         <span class="dt">col_names =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw">as.data.frame</span>()</code></pre></div>
<p>Note that <code>read.csv2()</code> treats semicolon “;” as the default separator and comma “,” as decimal, which is the standard in some regions. To change the separator to fx a comma you can adjust with the argument <code>sep = ","</code>. Refer to the documentation of <code>?read.table</code>.</p>
<p>After the two files have been loaded into R, check that the resulting data frames have been loaded correctly so that the first column in the otutable is a sample and not the OTU ID’s and that the first column in the metadata are the sample ID’s. It is highly recommended to read the documentation of <code><a href="../reference/amp_load.html">amp_load()</a></code> when loading your own data.</p>
<p>Now combine the data using <code><a href="../reference/amp_load.html">amp_load()</a></code>, which checks the data and combines it into one <code>ampvis2</code> object, making it easier to manipulate, filter and subset all elements of the data at once for analysis:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ampvis2)
d &lt;-<span class="st"> </span><span class="kw"><a href="../reference/amp_load.html">amp_load</a></span>(<span class="dt">otutable =</span> myotutable,
              <span class="dt">metadata =</span> mymetadata)</code></pre></div>
<p>The individual dataframes in the list can be explored with <code>View(d$metadata)</code>.</p>
</div>
<div id="filtering-and-subsetting" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-and-subsetting" class="anchor"></a>Filtering and subsetting</h2>
<p>With the ampvis2 package comes a large example data set with 573 samples taken from the activated sludge from 55 Danish Wastewater Treatment Plants in the period 2006-2013, which can be loaded with <code>data("MiDAS")</code>. The loaded data can be subsetted based on variables in the metadata using the <code><a href="../reference/amp_subset_samples.html">amp_subset_samples()</a></code> function, which can then be stored as a new object and analysed separately:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"MiDAS"</span>)
MiDAS</code></pre></div>
<pre><code>## ampvis2 object with 4 elements.
## Summary of OTU table:
##      Samples         OTUs  Total#Reads    Min#Reads    Max#Reads 
##          573        14791     18398563        13306        46071 
## Median#Reads    Avg#Reads 
##        31758     32109.18 
## 
## Assigned taxonomy:
##       Kingdom        Phylum         Class         Order        Family 
##   14791(100%) 14283(96.57%) 12507(84.56%) 11184(75.61%)  9685(65.48%) 
##         Genus       Species 
##  7105(48.04%)     27(0.18%) 
## 
## Metadata variables:
##  SampleID Plant Year Period Date DeltaDays LIB_ID Latitude Longitude RawData</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MiDASsubset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/amp_subset_samples.html">amp_subset_samples</a></span>(MiDAS, Plant %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">"Aalborg West"</span>, <span class="st">"Aalborg East"</span>))</code></pre></div>
<pre><code>## 522 samples and 6569 OTUs have been filtered 
## Before: 573 samples and 14791 OTUs
## After: 51 samples and 8222 OTUs</code></pre>
<p>or for a more complex subset, you can subset based on two or more variables using “&amp;” to separate the conditions, or simply use the function more than once. The “!” (logical NOT operator) can be thought of as “except” and is useful to remove fx outliers. Furthermore, the <code>minreads = 10000</code> argument removes any sample(s) with total amount of reads below the chosen threshold:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MiDASsubset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/amp_subset_samples.html">amp_subset_samples</a></span>(MiDAS, Plant %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">"Aalborg West"</span>, <span class="st">"Aalborg East"</span>) &amp;<span class="st"> </span>!SampleID %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">"16SAMP-749"</span>), <span class="dt">minreads =</span> <span class="dv">10000</span>)</code></pre></div>
<pre><code>## 523 samples and 6602 OTUs have been filtered 
## Before: 573 samples and 14791 OTUs
## After: 50 samples and 8189 OTUs</code></pre>
<p>The <code><a href="../reference/amp_subset_taxa.html">amp_subset_taxa()</a></code> function instead subsets based on the taxonomy, where you simply provide a vector with the taxa you are interested in, separated by a comma:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MiD_Chloroflexi_Actinobacteria &lt;-<span class="st"> </span><span class="kw"><a href="../reference/amp_subset_taxa.html">amp_subset_taxa</a></span>(MiDAS, <span class="dt">tax_vector=</span><span class="kw">c</span>(<span class="st">"p__Chloroflexi"</span>, <span class="st">"p__Actinobacteria"</span>))</code></pre></div>
<p>The taxonomic rank is indicated by fx “p__" for phylum and “g__" for genus etc, followed by the name of the taxon (case-sensitive, first letter almost always capital). To filter individual OTUs simply provide the OTU name(s) as-is in a vector, fx <code>c("OTU_1206")</code>.</p>
</div>
<div id="heatmap" class="section level1">
<h1 class="hasAnchor">
<a href="#heatmap" class="anchor"></a>Heatmap</h1>
<p>All ampvis2 plots are generated using the <a href="http://ggplot2.tidyverse.org/">ggplot2</a> package. You can change the look of the plots to better suit your needs, add more layers to the plots and use other ggplot2 functions in combination with ampvis plots if needed. Refer to the <a href="http://ggplot2.tidyverse.org/reference/index.html">ggplot2 documentation</a> for more information. <code><a href="../reference/amp_heatmap.html">amp_heatmap()</a></code> by default aggregates to phylum level and shows the top 10 phyla, ordered by mean read abundance across all samples:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/amp_heatmap.html">amp_heatmap</a></span>(MiDASsubset, <span class="dt">group_by =</span> <span class="st">"Plant"</span>)</code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<p>There are many arguments you can use to suit your needs, for a full list see the reference (click the function names to go to its reference page). For example, you can manually select the level at which to aggregate, how many to show, add additional higher level taxonomic information, group the samples differently by the metadata, hide the values, change the colors and scaling, and much more. You can also adjust the text labels for better readability or adjust the positioning of the legend (adjusting ggplot2 plots is always done with “+” after the actual ampvis2 function):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/amp_heatmap.html">amp_heatmap</a></span>(MiDASsubset,
            <span class="dt">group_by =</span> <span class="kw">c</span>(<span class="st">"Plant"</span>, <span class="st">"Year"</span>),
            <span class="dt">tax_aggregate =</span> <span class="st">"Genus"</span>,
            <span class="dt">tax_add =</span> <span class="st">"Phylum"</span>,
            <span class="dt">tax_show =</span> <span class="dv">25</span>,
            <span class="dt">color_vector =</span> <span class="kw">c</span>(<span class="st">"white"</span>, <span class="st">"red"</span>),
            <span class="dt">plot_colorscale =</span> <span class="st">"log10"</span>,
            <span class="dt">plot_values =</span> <span class="ot">FALSE</span>) +
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">45</span>, <span class="dt">size=</span><span class="dv">10</span>, <span class="dt">vjust =</span> <span class="dv">1</span>),
        <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">8</span>),
        <span class="dt">legend.position=</span><span class="st">"right"</span>)</code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
</div>
<div id="boxplot" class="section level1">
<h1 class="hasAnchor">
<a href="#boxplot" class="anchor"></a>Boxplot</h1>
<p><code><a href="../reference/amp_boxplot.html">amp_boxplot()</a></code> generates boxplots, again ordered by mean read abundance across all samples:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/amp_boxplot.html">amp_boxplot</a></span>(MiDASsubset)</code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>The arguments you can provide are similar to those used in <code><a href="../reference/amp_heatmap.html">amp_heatmap()</a></code> and other ampvis2 functions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/amp_boxplot.html">amp_boxplot</a></span>(MiDASsubset,
            <span class="dt">group_by =</span> <span class="st">"Period"</span>,
            <span class="dt">tax_show =</span> <span class="dv">5</span>,
            <span class="dt">tax_add =</span> <span class="st">"Phylum"</span>)</code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
</div>
<div id="ordination" class="section level1">
<h1 class="hasAnchor">
<a href="#ordination" class="anchor"></a>Ordination</h1>
<p>The <code><a href="../reference/amp_ordinate.html">amp_ordinate()</a></code> function has been expanded to support 7 different ordination methods, various data transformations and interactive plots by using <a href="https://plot.ly/">Plotly</a>. By default any OTU with an abundance no higher than 0.1% in any sample is removed, which drastically improves the calculation time. You can of course adjust this threshold manually by changing the <code>filter_species = 0.1</code> argument. Other than this, there are only four main arguments that are involved in the actual calculations, the rest are just various plotting features. These four are <code>type = ""</code>, <code>transform = ""</code>, <code>distmeasure = ""</code>, and lastly <code>constrain = ""</code> for constrained ordination (only used in Redundancy Analysis (RDA) or Canonical Correspondence Analysis (CCA)).</p>
<p>When analysing microbial community composition data it is recommended to use the Hellinger transformation (see <a href="https://doi.org/10.1007/s004420100716">Legendre et al, 2001</a> or <a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Numerical Ecology</a> for details) for most types of ordination methods except the distance-based ordination methods (Principal Coordinates Analysis (PCoA) and non-Metric Multidimensional Scaling (nMDS)), where you also have to select a distance measure manually by the <code>distmeasure = ""</code> argument, for example Bray-Curtis dissimilarities:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span>(MiDASsubset, 
             <span class="dt">type =</span> <span class="st">"pcoa"</span>,
             <span class="dt">distmeasure =</span> <span class="st">"bray"</span>,
             <span class="dt">sample_color_by =</span> <span class="st">"Plant"</span>,
             <span class="dt">sample_colorframe =</span> <span class="ot">TRUE</span>,
             <span class="dt">sample_colorframe_label =</span> <span class="st">"Plant"</span>) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"blank"</span>)</code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>And lastly an example of constrained ordination by Canonical Correspondence Analysis (CCA), which in this case reveals how samples taken at different seasonal periods (<code>constrain = "Period"</code>) of the year can explain the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ordinationresult &lt;-<span class="st"> </span><span class="kw"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span>(MiDASsubset, 
             <span class="dt">type =</span> <span class="st">"CCA"</span>,
             <span class="dt">constrain =</span> <span class="st">"Period"</span>,
             <span class="dt">transform =</span> <span class="st">"Hellinger"</span>,
             <span class="dt">sample_color_by =</span> <span class="st">"Period"</span>,
             <span class="dt">sample_shape_by =</span> <span class="st">"Plant"</span>,
             <span class="dt">sample_colorframe =</span> <span class="ot">TRUE</span>,
             <span class="dt">sample_colorframe_label =</span> <span class="st">"Period"</span>,
             <span class="dt">detailed_output =</span> <span class="ot">TRUE</span>)
ordinationresult$plot</code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<p>Notice that the plot can be saved as a more detailed object (<code>detailed_output = TRUE</code>), so additional details about the ordination result can be obtained for evaluation, fx a screeplot by <code>ordinationresult$screeplot</code> or some more raw data by <code>ordinationresult$model</code>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#loading-data">Loading data</a></li>
      <li><a href="#filtering-and-subsetting">Filtering and subsetting</a></li>
      <li><a href="#heatmap">Heatmap</a></li>
      <li><a href="#boxplot">Boxplot</a></li>
      <li><a href="#ordination">Ordination</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Mads Albertsen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
