<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load data for ampvis2 functions — amp_load • ampvis2</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Load data for ampvis2 functions — amp_load" />
<meta property="og:description" content="This function reads an OTU-table and corresponding sample metadata, and returns a list for use in all ampvis2 functions. It is therefore required to load data with amp_load before any other ampvis2 functions can be used." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-2');
</script>


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ampvis2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/ampvis2.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/faq.html">FAQ</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/MadsAlbertsen/ampvis2">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load data for ampvis2 functions</h1>
    <small class="dont-index">Source: <a href='https://github.com/MadsAlbertsen/ampvis2/blob/master/R/amp_load.R'><code>R/amp_load.R</code></a></small>
    <div class="hidden name"><code>amp_load.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function reads an OTU-table and corresponding sample metadata, and returns a list for use in all ampvis2 functions. It is therefore required to load data with <code>amp_load</code> before any other ampvis2 functions can be used.</p>
    </div>

    <pre class="usage"><span class='fu'>amp_load</span><span class='op'>(</span>
  <span class='va'>otutable</span>,
  metadata <span class='op'>=</span> <span class='cn'>NULL</span>,
  taxonomy <span class='op'>=</span> <span class='cn'>NULL</span>,
  fasta <span class='op'>=</span> <span class='cn'>NULL</span>,
  tree <span class='op'>=</span> <span class='cn'>NULL</span>,
  pruneSingletons <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>otutable</th>
      <td><p>(<em>required</em>) OTU-table with the read counts of all OTU's. Rows are OTU's, columns are samples, otherwise you must transpose. The taxonomy of the OTU's can be placed anywhere in the table and will be extracted by name (Kingdom/Domain -&gt; Species). Can be a data frame, matrix, or path to a delimited text file or excel file which will be read using either <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>fread</a></code> or <code><a href='https://readxl.tidyverse.org/reference/read_excel.html'>read_excel</a></code>, respectively.</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>(<em>recommended</em>) Sample metadata with any information about the samples. The first column must contain sample ID's matching those in the otutable. If none provided, dummy metadata will be created. Can be a data frame, matrix, or path to a delimited text file or excel file which will be read using either <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>fread</a></code> or <code><a href='https://readxl.tidyverse.org/reference/read_excel.html'>read_excel</a></code>, respectively. (<em>default:</em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>taxonomy</th>
      <td><p>(<em>recommended</em>) Taxonomy table where rows are OTU's and columns are up to 7 levels of taxonomy named Kingdom/Domain-&gt;Species. If taxonomy is also present in otutable, it will be discarded and only this will be used. Can be a data frame, matrix, or path to a delimited text file or excel file which will be read using either <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>fread</a></code> or <code><a href='https://readxl.tidyverse.org/reference/read_excel.html'>read_excel</a></code>, respectively. (<em>default:</em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>fasta</th>
      <td><p>(<em>optional</em>) Path to a FASTA file with reference sequences for all OTU's in the OTU-table. (<em>default:</em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>tree</th>
      <td><p>(<em>optional</em>) Path to a phylogenetic tree file which will be read using <code><a href='https://rdrr.io/pkg/ape/man/read.tree.html'>read.tree</a></code>, or an object of class <code>"phylo"</code>. (<em>default:</em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>pruneSingletons</th>
      <td><p>(<em>logical</em>) Remove OTU's only observed once in all samples. (<em>default:</em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(<em>optional</em>) Additional arguments are passed on to any of the file reader functions used.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of class <code>"ampvis2"</code> with 3 to 5 elements.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>amp_load</code> function validates and corrects the provided data frames in different ways to make it suitable for the rest of the ampvis2 functions. It is important that the provided data frames match the requirements as described in the following sections to work properly.</p>
    <h2 class="hasAnchor" id="the-otu-table"><a class="anchor" href="#the-otu-table"></a>The OTU-table</h2>

    

<p>The OTU-table contains information about the OTUs, their read counts in each sample, and optionally their assigned taxonomy. The provided OTU-table must be a data frame with the following requirements:</p>
<ul>
<li><p>The rows are OTU IDs and the columns are samples.</p></li>
<li><p>The last 7 columns are optionally the corresponding taxonomy assigned to the OTUs, named <code>"Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"</code>.</p></li>
<li><p>The OTU ID's are expected to be in either the row names of the data frame or in a column called "OTU", "ASV", or "#OTU ID". Otherwise the function will stop with a message.</p></li>
<li><p>The column names of the data frame are the sample IDs, exactly matching those in the metadata, (and taxonomy columns named <code>Kingdom</code> -&gt; <code>Species</code> if present, of course).</p></li>
<li><p>Generally avoid special characters and spaces in row- and column names.</p></li>
</ul>

<p>A minimal example is available with <code><a href='https://rdrr.io/r/utils/data.html'>data("example_otutable")</a></code>.</p>
    <h2 class="hasAnchor" id="the-metadata"><a class="anchor" href="#the-metadata"></a>The metadata</h2>

    

<p>The metadata contains additional information about the samples, for example where each sample was taken, date, pH, treatment etc, which is used to compare and group the samples during analysis. The amount of information in the metadata is unlimited, it can contain any number of columns (variables), however there are a few requirements:</p>
<ul>
<li><p>The sample IDs must be in the first column. These sample IDs must match exactly to those in the OTU-table.</p></li>
<li><p>Column classes matter, categorical variables should be loaded either <code><a href='https://rdrr.io/r/base/character.html'>as.character()</a></code> or <code><a href='https://rdrr.io/r/base/factor.html'>as.factor()</a></code>, and continuous variables <code><a href='https://rdrr.io/r/base/numeric.html'>as.numeric()</a></code>. See below.</p></li>
<li><p>Generally avoid special characters and spaces in row- and column names.</p></li>
</ul>

<p>If for example a column is named "Year" and the entries are simply entered as numbers (2011, 2012, 2013 etc), then R will automatically consider these as numerical values (<code><a href='https://rdrr.io/r/base/numeric.html'>as.numeric()</a></code>) and therefore the column as a continuous variable, while it is a categorical variable and should be loaded <code><a href='https://rdrr.io/r/base/factor.html'>as.factor()</a></code> or <code><a href='https://rdrr.io/r/base/character.html'>as.character()</a></code> instead. This has consequences for the analysis as R treats them differently. Therefore either use the <code>colClasses = </code> argument when loading a csv file or <code>col_types = </code> when loading an excel file, or manually adjust the column classes afterwards with fx <code>metadata$Year &lt;- as.character(metadata$Year)</code>.</p>
<p>The <code>amp_load</code> function will automatically use the sample IDs in the first column as row names, but it is important to also have an actual column with sample IDs, so it's possible to fx group by that column during analysis. Any unmatched samples between the otutable and metadata will be removed with a warning.</p>
<p>A minimal example is available with <code><a href='https://rdrr.io/r/utils/data.html'>data("example_metadata")</a></code>.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>amp_load</code>, <code><a href='amp_subset_samples.html'>amp_subset_samples</a></code>, <code><a href='amp_subset_taxa.html'>amp_subset_taxa</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Kasper Skytte Andersen <a href='mailto:ksa@bio.aau.dk'>ksa@bio.aau.dk</a></p>
<p>Mads Albertsen <a href='mailto:MadsAlbertsen85@gmail.com'>MadsAlbertsen85@gmail.com</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mads Albertsen, Kasper Skytte Andersen, Rasmus Hansen Kirkegaard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'b0ba3a3eb751685982ccf010113f380a',
    indexName: 'ampvis2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  

  </body>
</html>


