<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Subset ampvis2 objects based on sample metadata — amp_subset_samples • ampvis2</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Subset ampvis2 objects based on sample metadata — amp_subset_samples" />
<meta property="og:description" content="Subsets the data in ampvis2 objects based on metadata and returns the subsetted object." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-2');
</script>


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ampvis2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/ampvis2.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/faq.html">FAQ</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/MadsAlbertsen/ampvis2">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subset ampvis2 objects based on sample metadata</h1>
    <small class="dont-index">Source: <a href='https://github.com/MadsAlbertsen/ampvis2/blob/master/R/amp_subset_samples.R'><code>R/amp_subset_samples.R</code></a></small>
    <div class="hidden name"><code>amp_subset_samples.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Subsets the data in ampvis2 objects based on metadata and returns the subsetted object.</p>
    </div>

    <pre class="usage"><span class='fu'>amp_subset_samples</span><span class='op'>(</span>
  <span class='va'>data</span>,
  <span class='va'>...</span>,
  minreads <span class='op'>=</span> <span class='fl'>0</span>,
  rarefy <span class='op'>=</span> <span class='cn'>NULL</span>,
  normalise <span class='op'>=</span> <span class='cn'>FALSE</span>,
  removeAbsents <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>(<em>required</em>) Data list as loaded with <code><a href='amp_load.html'>amp_load</a></code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Logical expression indicating elements or rows to keep in the metadata. Missing values are treated as <code>FALSE</code>. Passed directly to <code><a href='https://dplyr.tidyverse.org/reference/filter.html'>filter</a></code>.</p></td>
    </tr>
    <tr>
      <th>minreads</th>
      <td><p>Minimum number of reads pr. sample. Samples below this value will be removed <em>initially</em>. (<em>default:</em> <code>0</code>)</p></td>
    </tr>
    <tr>
      <th>rarefy</th>
      <td><p>Rarefy species richness to this value by using <code><a href='https://rdrr.io/pkg/vegan/man/rarefy.html'>rrarefy</a></code>. This is done <em>initially</em>, but <em>after</em> filtering based on the <code>minreads</code> value, if set. (<em>default:</em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>normalise</th>
      <td><p>(<em>logical</em>) Normalise the OTU read counts to 100 (ie percent) per sample <em>BEFORE</em> the subset. (<em>default:</em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>removeAbsents</th>
      <td><p>(<em>logical</em>) Whether to remove OTU's that may have 0 read abundance in all samples after the subset. (<em>default:</em> <code>TRUE</code>)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A modifed ampvis2 object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The subset is performed on the metadata by <code><a href='https://rdrr.io/r/base/subset.html'>subset()</a></code> and the abundance- and taxonomy tables are then adjusted accordingly.</p>
    <h2 class="hasAnchor" id="preserving-relative-abundances-in-a-subset-of-larger-data"><a class="anchor" href="#preserving-relative-abundances-in-a-subset-of-larger-data"></a>Preserving relative abundances in a subset of larger data</h2>

    

<p>By default the raw read counts in the abundance matrix are normalised (transformed to percentages) by some plotting functions automatically (for example <code><a href='amp_heatmap.html'>amp_heatmap</a></code>, <code><a href='amp_timeseries.html'>amp_timeseries</a></code>, and more). This means that the relative abundances shown will be calculated based on the remaining taxa after the subset, not including the removed taxa, if any. To circumvent this, set <code>normalise = TRUE</code> when subsetting with the <code><a href='amp_subset_taxa.html'>amp_subset_taxa</a></code> and <code>amp_subset_samples</code> functions, and then set <code>normalise = FALSE</code> in the plotting function. This will transform the OTU counts to relative abundances BEFORE the subset, and setting <code>normalise = FALSE</code> will skip the transformation in the plotting function, see the example below.</p>
<pre>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"MiDAS"</span><span class='op'>)</span>
<span class='va'>subsettedData</span> <span class='op'>&lt;-</span> <span class='fu'>amp_subset_samples</span><span class='op'>(</span><span class='va'>MiDAS</span>,
                                    <span class='va'>Plant</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Aalborg West"</span>, <span class='st'>"Aalborg East"</span><span class='op'>)</span>,
                                    normalise <span class='op'>=</span> <span class='cn'>TRUE</span>
                                    <span class='op'>)</span>
<span class='fu'><a href='amp_heatmap.html'>amp_heatmap</a></span><span class='op'>(</span><span class='va'>subsettedData</span>,
            group_by <span class='op'>=</span> <span class='st'>"Plant"</span>,
            tax_aggregate <span class='op'>=</span> <span class='st'>"Phylum"</span>,
            tax_add <span class='op'>=</span> <span class='st'>"Genus"</span>,
            normalise <span class='op'>=</span> <span class='cn'>FALSE</span>
            <span class='op'>)</span>
</pre>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>McMurdie, P.J. &amp; Holmes, S. (2014). Waste not, want not: Why
rarefying microbiome data is inadmissible. <em>PLoS Comput Biol</em>
<strong>10(4):</strong> e1003531. DOI:<code>10.1371/journal.pcbi.1003531</code></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='amp_load.html'>amp_load</a></code>, <code><a href='amp_subset_taxa.html'>amp_subset_taxa</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Kasper Skytte Andersen <a href='mailto:ksa@bio.aau.dk'>ksa@bio.aau.dk</a></p>
<p>Mads Albertsen <a href='mailto:MadsAlbertsen85@gmail.com'>MadsAlbertsen85@gmail.com</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Load example data</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"MiDAS"</span><span class='op'>)</span>

<span class='co'># Show a short summary about the data by simply typing the name of the object in the console</span>
<span class='va'>MiDAS</span>
</div><div class='output co'>#&gt; ampvis2 object with 5 elements. <span style='text-decoration: underline;'>
#&gt; Summary of OTU table:
#&gt; </span><span>     Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads 
#&gt;          658        14969     20890850        10480        46264        31800 
#&gt;    Avg#Reads 
#&gt;     31749.01 
#&gt; </span><span style='text-decoration: underline;'>
#&gt; Assigned taxonomy:
#&gt; </span><span>      Kingdom        Phylum         Class         Order        Family 
#&gt;   14969(100%) 14477(96.71%) 12737(85.09%) 11470(76.63%)  9841(65.74%) 
#&gt;         Genus       Species 
#&gt;   7380(49.3%)     28(0.19%) 
#&gt; </span><span style='text-decoration: underline;'>
#&gt; Metadata variables:</span><span> 5 
#&gt;  SampleID, Plant, Date, Year, Period</div><div class='input'>
<span class='co'># Keep only samples containing Aalborg West or East in the Plant column</span>
<span class='va'>MiDASsubset</span> <span class='op'>&lt;-</span> <span class='fu'>amp_subset_samples</span><span class='op'>(</span><span class='va'>MiDAS</span>, <span class='va'>Plant</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Aalborg West"</span>, <span class='st'>"Aalborg East"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>590 samples and 5512 OTUs have been filtered </span>
#&gt; <span class='message'>Before: 658 samples and 14969 OTUs</span>
#&gt; <span class='message'>After: 68 samples and 9457 OTUs</span></div><div class='input'>
<span class='co'># Summary</span>
<span class='va'>MiDASsubset</span>
</div><div class='output co'>#&gt; ampvis2 object with 5 elements. </span><span style='text-decoration: underline;'>
#&gt; Summary of OTU table:
#&gt; </span><span>     Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads 
#&gt;           68         9457      2072678        17772        44326      30962.5 
#&gt;    Avg#Reads 
#&gt;     30480.56 
#&gt; </span><span style='text-decoration: underline;'>
#&gt; Assigned taxonomy:
#&gt; </span><span>     Kingdom       Phylum        Class        Order       Family        Genus 
#&gt;   9457(100%) 9240(97.71%) 8305(87.82%) 7665(81.05%) 6767(71.56%) 5244(55.45%) 
#&gt;      Species 
#&gt;    23(0.24%) 
#&gt; </span><span style='text-decoration: underline;'>
#&gt; Metadata variables:</span><span> 5 
#&gt;  SampleID, Plant, Date, Year, Period</div><div class='input'>
<span class='co'># Keep only samples containing Aalborg West or East in the Plant column</span>
<span class='co'># and remove the sample "16SAMP-749". Remove any sample(s) with less than 20000 total reads</span>
<span class='va'>MiDASsubset2</span> <span class='op'>&lt;-</span> <span class='fu'>amp_subset_samples</span><span class='op'>(</span><span class='va'>MiDAS</span>,
  <span class='va'>Plant</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Aalborg West"</span>, <span class='st'>"Aalborg East"</span><span class='op'>)</span> <span class='op'>&amp;</span> <span class='op'>!</span><span class='va'>SampleID</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"16SAMP-749"</span><span class='op'>)</span>,
  minreads <span class='op'>=</span> <span class='fl'>20000</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>594 samples and 5601 OTUs have been filtered </span>
#&gt; <span class='message'>Before: 658 samples and 14969 OTUs</span>
#&gt; <span class='message'>After: 64 samples and 9368 OTUs</span></div><div class='input'>
<span class='co'># Summary</span>
<span class='va'>MiDASsubset2</span>
</div><div class='output co'>#&gt; ampvis2 object with 5 elements. </span><span style='text-decoration: underline;'>
#&gt; Summary of OTU table:
#&gt; </span><span>     Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads 
#&gt;           64         9368      1987574        21472        44326      31413.5 
#&gt;    Avg#Reads 
#&gt;     31055.84 
#&gt; </span><span style='text-decoration: underline;'>
#&gt; Assigned taxonomy:
#&gt; </span><span>     Kingdom       Phylum        Class        Order       Family        Genus 
#&gt;   9368(100%) 9154(97.72%) 8222(87.77%) 7589(81.01%) 6699(71.51%) 5194(55.44%) 
#&gt;      Species 
#&gt;    23(0.25%) 
#&gt; </span><span style='text-decoration: underline;'>
#&gt; Metadata variables:</span><span> 5 
#&gt;  SampleID, Plant, Date, Year, Period</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Mads Albertsen, Kasper Skytte Andersen, Rasmus Hansen Kirkegaard.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9000.</p></p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'b0ba3a3eb751685982ccf010113f380a',
    indexName: 'ampvis2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  

  </body>
</html>


