<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to ampvis2 • ampvis2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Introduction to ampvis2">
<meta property="og:description" content="ampvis2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ampvis2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.7.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/ampvis2.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/faq.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MadsAlbertsen/ampvis2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ampvis2_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to ampvis2</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MadsAlbertsen/ampvis2/blob/master/vignettes/ampvis2.Rmd"><code>vignettes/ampvis2.Rmd</code></a></small>
      <div class="hidden name"><code>ampvis2.Rmd</code></div>

    </div>

    
    
<p>This is a short guide to the basics of ampvis2, how to load data, as well as a few basic visualisation functions using example data. A complete explanation of all functions, their purpose, different arguments and examples can be found in the <a href="https://madsalbertsen.github.io/ampvis2/reference/index.html">Functions</a> tab. First, install ampvis2 as described on the front page.</p>
<div id="loading-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading data</h2>
<p>ampvis2 imports data directly from any of the commonly used <a href="http://qiime.org/">QIIME</a>, <a href="https://www.mothur.org/">mothur</a>, and <a href="http://www.drive5.com/uparse/">UPARSE</a> amplicon pipelines. Both QIIME and mothur (through the <a href="https://www.mothur.org/wiki/Make.biom">make.biom</a> script) can export data in the <a href="http://biom-format.org/">BIOM format</a> and can be loaded with the <a href="https://madsalbertsen.github.io/ampvis2/reference/amp_import_biom.html">amp_import_biom</a> function. Data from the UPARSE pipeline can be imported by the <a href="https://madsalbertsen.github.io/ampvis2/reference/amp_import_usearch.html">amp_import_uparse</a> function. The data is then loaded into a single <code>ampvis2</code> object by the <code><a href="../reference/amp_load.html">amp_load()</a></code> function, which results in a list of data frames, see the below illustration of the data structure and typical workflow of ampvis2:</p>
<p><img src="workflow.png"></p>
<p>To demonstrate how to load data from raw CSV files download and unzip this <a href="https://doi.org/10.6084/m9.figshare.6139352">minimal example data</a>. The otutable.csv and metadata.csv files can then be imported into R by an appropriate read function:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">myotutable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"example_otutable.csv"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">mymetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"example_metadata.csv"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Note that there are multiple types of CSV files. To change the separator to fx a comma you can adjust with the argument <code>sep = ","</code>. Refer to the documentation of <a href="https://www.rdocumentation.org/packages/utils/versions/3.4.1/topics/read.table">read.table</a>. To load excel files use the <code><a href="https://readxl.tidyverse.org/reference/read_excel.html">readxl::read_excel</a></code> function instead.</p>
<p>After the two files have been loaded into R, check that the resulting data frames have been loaded correctly (click them in the working environment to the right under “Data”), so that either the rownames or a column named “OTU” contains the OTU ID’s in the loaded otutable, and that the first column in the metadata are the matching sample ID’s. It is highly recommended to read the documentation of <code><a href="../reference/amp_load.html">amp_load()</a></code> before loading data.</p>
<p>Now combine the data using <code><a href="../reference/amp_load.html">amp_load()</a></code>, which checks the data and combines it into a single <code>ampvis2</code> object, making it easier to manipulate, filter and subset all elements of the data at once for analysis:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://madsalbertsen.github.io/ampvis2">ampvis2</a></span><span class="op">)</span>
<span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_load.html">amp_load</a></span><span class="op">(</span>otutable <span class="op">=</span> <span class="va">myotutable</span>,
              metadata <span class="op">=</span> <span class="va">mymetadata</span><span class="op">)</span></code></pre></div>
<p>The individual dataframes in the list can be explored with <code><a href="https://rdrr.io/r/utils/View.html">View(d$metadata)</a></code>.</p>
</div>
<div id="filtering-and-subsetting" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-and-subsetting" class="anchor"></a>Filtering and subsetting</h2>
<p>With the ampvis2 package comes a large example data set with 573 samples taken from the activated sludge from 55 Danish Wastewater Treatment Plants in the period 2006-2013, which can be loaded with <code><a href="https://rdrr.io/r/utils/data.html">data("MiDAS")</a></code>. Simply typing the name of any ampvis2 object in the console will show a short summary of the data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"MiDAS"</span><span class="op">)</span>
<span class="va">MiDAS</span></code></pre></div>
<pre><code>## ampvis2 object with 5 elements. 
## Summary of OTU table:
##      Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads 
##          658        14969     20890850        10480        46264        31800 
##    Avg#Reads 
##     31749.01 
## 
## Assigned taxonomy:
##       Kingdom        Phylum         Class         Order        Family 
##   14969(100%) 14477(96.71%) 12737(85.09%) 11470(76.63%)  9841(65.74%) 
##         Genus       Species 
##   7380(49.3%)     28(0.19%) 
## 
## Metadata variables: 5 
##  SampleID, Plant, Date, Year, Period</code></pre>
<p>If you have loaded the raw DNA sequences of the OTUs from a FASTA file (with the <code>fasta =</code> argument) you can also get a short summary by typing the name of the ampvis2 object followed by <code>$refseq</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MiDAS</span><span class="op">$</span><span class="va">refseq</span></code></pre></div>
<pre><code>## 14969 DNA sequences in binary format stored in a list.
## 
## Mean sequence length: 472.922 
##    Shortest sequence: 425 
##     Longest sequence: 525 
## 
## Labels:
## OTU_1
## OTU_2
## OTU_3
## OTU_4
## OTU_5
## OTU_6
## ...
## 
## Base composition:
##     a     c     g     t 
## 0.261 0.225 0.319 0.194 
## (Total: 7.08 Mb)</code></pre>
<p>The loaded data can be subsetted based on variables in the metadata using the <code><a href="../reference/amp_subset_samples.html">amp_subset_samples()</a></code> function, which can then be stored as a new object and analysed separately:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MiDASsubset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_subset_samples.html">amp_subset_samples</a></span><span class="op">(</span><span class="va">MiDAS</span>, <span class="va">Plant</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Aalborg West"</span>, <span class="st">"Aalborg East"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 590 samples and 5512 OTUs have been filtered 
## Before: 658 samples and 14969 OTUs
## After: 68 samples and 9457 OTUs</code></pre>
<p>or for a more complex subset, you can subset based on two or more variables using “&amp;” to separate the conditions, or simply use the function more than once. The “!” (logical NOT operator) can be thought of as “except” and is useful to remove fx outliers. Furthermore, the <code>minreads = 10000</code> argument removes any sample(s) with total amount of reads below the chosen threshold:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MiDASsubset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_subset_samples.html">amp_subset_samples</a></span><span class="op">(</span><span class="va">MiDAS</span>, <span class="va">Plant</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Aalborg West"</span>, <span class="st">"Aalborg East"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">SampleID</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"16SAMP-749"</span><span class="op">)</span>, minreads <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></code></pre></div>
<pre><code>## 591 samples and 5539 OTUs have been filtered 
## Before: 658 samples and 14969 OTUs
## After: 67 samples and 9430 OTUs</code></pre>
<p>The <code><a href="../reference/amp_subset_taxa.html">amp_subset_taxa()</a></code> function instead subsets based on the taxonomy, where you simply provide a vector with the taxa you are interested in, separated by a comma:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MiDAS_Chloroflexi_Actinobacteria</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_subset_taxa.html">amp_subset_taxa</a></span><span class="op">(</span><span class="va">MiDAS</span>, tax_vector<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"p__Chloroflexi"</span>, <span class="st">"p__Actinobacteria"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 12245 OTUs have been filtered 
## Before: 14969 OTUs
## After: 2724 OTUs</code></pre>
<p>The taxonomic rank is indicated by fx "p__" for phylum and "g__" for genus etc, followed by the name of the taxon (case-sensitive, first letter almost always capital). To filter individual OTUs simply provide the OTU name(s) as-is in a vector, fx <code><a href="https://rdrr.io/r/base/c.html">c("OTU_1206")</a></code>.</p>
</div>
<div id="heatmap" class="section level1">
<h1 class="hasAnchor">
<a href="#heatmap" class="anchor"></a>Heatmap</h1>
<p>All ampvis2 plots are generated using the <a href="http://ggplot2.tidyverse.org/">ggplot2</a> package. You can change the look of the plots to better suit your needs, add more layers to the plots and use other ggplot2 functions in combination with ampvis plots if needed. Refer to the <a href="http://ggplot2.tidyverse.org/reference/index.html">ggplot2 documentation</a> for more information. <code><a href="../reference/amp_heatmap.html">amp_heatmap()</a></code> by default aggregates to phylum level and shows the top 10 phyla, ordered by mean read abundance across all samples:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/amp_heatmap.html">amp_heatmap</a></span><span class="op">(</span><span class="va">MiDASsubset</span>, group_by <span class="op">=</span> <span class="st">"Plant"</span><span class="op">)</span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>There are many arguments you can use to suit your needs, for a full list see the reference (click the function names to go to its reference page). For example, you can manually select the level at which to aggregate, how many to show, add additional higher level taxonomic information, group the samples differently by the metadata, hide the values, change the colors and scaling, and much more. You can also adjust the text labels for better readability or adjust the positioning of the legend (adjusting ggplot2 plots is always done with “+” after the actual ampvis2 function):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/amp_heatmap.html">amp_heatmap</a></span><span class="op">(</span><span class="va">MiDASsubset</span>,
            group_by <span class="op">=</span> <span class="st">"Plant"</span>,
            facet_by <span class="op">=</span> <span class="st">"Year"</span>,
            tax_aggregate <span class="op">=</span> <span class="st">"Genus"</span>,
            tax_add <span class="op">=</span> <span class="st">"Phylum"</span>,
            tax_show <span class="op">=</span> <span class="fl">25</span>,
            color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"darkred"</span><span class="op">)</span>,
            plot_colorscale <span class="op">=</span> <span class="st">"sqrt"</span>,
            plot_values <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, size<span class="op">=</span><span class="fl">10</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
        axis.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,
        legend.position<span class="op">=</span><span class="st">"right"</span><span class="op">)</span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="boxplot" class="section level1">
<h1 class="hasAnchor">
<a href="#boxplot" class="anchor"></a>Boxplot</h1>
<p><code><a href="../reference/amp_boxplot.html">amp_boxplot()</a></code> generates boxplots, again ordered by mean read abundance across all samples:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/amp_boxplot.html">amp_boxplot</a></span><span class="op">(</span><span class="va">MiDASsubset</span><span class="op">)</span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>The arguments you can provide are similar to those used in <code><a href="../reference/amp_heatmap.html">amp_heatmap()</a></code> and other ampvis2 functions:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/amp_boxplot.html">amp_boxplot</a></span><span class="op">(</span><span class="va">MiDASsubset</span>,
            group_by <span class="op">=</span> <span class="st">"Period"</span>,
            tax_show <span class="op">=</span> <span class="fl">5</span>,
            tax_add <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div id="ordination" class="section level1">
<h1 class="hasAnchor">
<a href="#ordination" class="anchor"></a>Ordination</h1>
<p>The <code><a href="../reference/amp_ordinate.html">amp_ordinate()</a></code> function has been expanded to support 7 different ordination methods, various data transformations and interactive plots by using <a href="https://plot.ly/">Plotly</a>. By default any OTU with an abundance no higher than 0.1% in any sample is removed, which drastically improves the calculation time. You can of course adjust this threshold manually by changing the <code>filter_species = 0.1</code> argument. Other than this, there are only four main arguments that are involved in the actual calculations, the rest are just various plotting features. These four are <code>type = ""</code>, <code>transform = ""</code>, <code>distmeasure = ""</code>, and lastly <code>constrain = ""</code> for constrained ordination (only used in Redundancy Analysis (RDA) or Canonical Correspondence Analysis (CCA)).</p>
<p>When analysing microbial community composition data it is recommended to use the Hellinger transformation (see <a href="https://doi.org/10.1007/s004420100716">Legendre et al, 2001</a> or <a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Numerical Ecology</a> for details) for most types of ordination methods except the distance-based ordination methods (Principal Coordinates Analysis (PCoA) and non-Metric Multidimensional Scaling (nMDS)), where you also have to select a distance measure manually by the <code>distmeasure = ""</code> argument, for example Bray-Curtis dissimilarities:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span><span class="op">(</span><span class="va">MiDASsubset</span>, 
             type <span class="op">=</span> <span class="st">"pcoa"</span>,
             distmeasure <span class="op">=</span> <span class="st">"bray"</span>,
             sample_color_by <span class="op">=</span> <span class="st">"Plant"</span>,
             sample_colorframe <span class="op">=</span> <span class="cn">TRUE</span>,
             sample_colorframe_label <span class="op">=</span> <span class="st">"Plant"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"blank"</span><span class="op">)</span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>It is also possible to track changes over time by setting the <code>sample_trajectory</code> and <code>sample_trajectory_group</code> arguments, to for example reveal temporal patterns per WWTP:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span><span class="op">(</span><span class="va">MiDASsubset</span>, 
             type <span class="op">=</span> <span class="st">"pcoa"</span>,
             distmeasure <span class="op">=</span> <span class="st">"bray"</span>,
             sample_color_by <span class="op">=</span> <span class="st">"Plant"</span>,
             sample_colorframe_label <span class="op">=</span> <span class="st">"Plant"</span>, 
             sample_trajectory <span class="op">=</span> <span class="st">"Date"</span>, 
             sample_trajectory_group <span class="op">=</span> <span class="st">"Plant"</span><span class="op">)</span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>And lastly an example of constrained ordination by Canonical Correspondence Analysis (CCA), which in this case reveals how samples taken at different seasonal periods (<code>constrain = "Period"</code>) of the year can explain the data:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ordinationresult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span><span class="op">(</span><span class="va">MiDASsubset</span>, 
             type <span class="op">=</span> <span class="st">"CCA"</span>,
             constrain <span class="op">=</span> <span class="st">"Period"</span>,
             transform <span class="op">=</span> <span class="st">"Hellinger"</span>,
             sample_color_by <span class="op">=</span> <span class="st">"Period"</span>,
             sample_shape_by <span class="op">=</span> <span class="st">"Plant"</span>,
             sample_colorframe <span class="op">=</span> <span class="cn">TRUE</span>,
             sample_colorframe_label <span class="op">=</span> <span class="st">"Period"</span>,
             detailed_output <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">ordinationresult</span><span class="op">$</span><span class="va">plot</span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Notice that the plot can be saved as a more detailed object (<code>detailed_output = TRUE</code>), so additional details about the ordination result can be obtained for evaluation, fx a screeplot by <code>ordinationresult$screeplot</code> or some more raw data by <code>ordinationresult$model</code>.</p>
</div>
<div id="read-more" class="section level1">
<h1 class="hasAnchor">
<a href="#read-more" class="anchor"></a>Read more</h1>
<p>There are numerous other functions to try out as well as complete documentation for each function highlighted above. Go explore in the <a href="../reference/index.html">Functions</a> tab.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kasper Skytte Andersen, Mads Albertsen, Rasmus Hansen Kirkegaard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'b0ba3a3eb751685982ccf010113f380a',
    indexName: 'ampvis2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
