<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to ampvis2 • ampvis2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Introduction to ampvis2">
<meta property="og:description" content="ampvis2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ampvis2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.7.34</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/ampvis2.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/faq.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kasperskytte/ampvis2" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ampvis2_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to ampvis2</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kasperskytte/ampvis2/blob/HEAD/vignettes/ampvis2.Rmd" class="external-link"><code>vignettes/ampvis2.Rmd</code></a></small>
      <div class="hidden name"><code>ampvis2.Rmd</code></div>

    </div>

    
    
<p>This is a short guide to the basics of <code>ampvis2</code>, how to load data, as well as a few basic visualisation functions using example data. A complete explanation of all functions, their purpose, different arguments and examples can be found in the <a href="https://kasperskytte.github.io/ampvis2/reference/index.html">Functions</a> tab.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>Install <code>ampvis2</code> as described on the <a href="https://kasperskytte.github.io/ampvis2/index.html">home page</a>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"kasperskytte/ampvis2"</span>, Ncpus <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively use the RStudio docker image with ampvis2 preinstalled as described on the <a href="https://kasperskytte.github.io/ampvis2/index.html">home page</a>.</p>
</div>
<div class="section level3">
<h3 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h3>
<p>The first step is always to import some data by using the <code>amp_load</code> function. The <code>amp_load</code> function is quite flexible and auto-detects several different formats, including csv or excel files, BIOM files, or even R objects directly. As a minimum, you always need to import an abundance table with counts of OTU’s/ASV’s per sample, the rest are optional, but any useful analysis involves loading at least some sample metadata and taxonomy too. Refer to the <a href="https://kasperskytte.github.io/ampvis2/reference/amp_load.html">amp_load</a> help page for details. A simple example would be:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kasperskytte.github.io/ampvis2">ampvis2</a></span><span class="op">)</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_load.html">amp_load</a></span><span class="op">(</span></span>
<span>  otutable <span class="op">=</span> <span class="st">"path/to/otutable.csv"</span>,</span>
<span>  metadata <span class="op">=</span> <span class="st">"path/to/samplemetadata.xlsx"</span>,</span>
<span>  taxonomy <span class="op">=</span> <span class="st">"path/to/taxonomy.csv"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>amp_load</code> function also supports importing data directly from any of the commonly used amplicon processing pipelines like <a href="http://qiime.org/" class="external-link">QIIME</a>, <a href="https://www.mothur.org/" class="external-link">mothur</a>, <a href="http://www.drive5.com/usearch/" class="external-link">USEARCH</a>, and <a href="https://benjjneb.github.io/dada2/" class="external-link">DADA2</a>. To import data in the <a href="http://biom-format.org/" class="external-link">BIOM format</a> from QIIME and mothur (through the <a href="https://www.mothur.org/wiki/Make.biom" class="external-link">make.biom</a> script), for example, simply supply the path to the <code>.biom</code> file to the <code>otutable</code> argument.</p>
<p>The data is then loaded and combined into a single <code>ampvis2</code> class object by the <code><a href="../reference/amp_load.html">amp_load()</a></code> function, which makes the workflow a bit simpler as everything is performed on a single object and can be manipulated by dedicated functions. The data structure used in <code>ampvis2</code> is inspired by that of the <a href="https://joey711.github.io/phyloseq/index.html" class="external-link">phyloseq</a> R package, but instead of being an abstract S4 class object it’s a simple <code>list</code> of data frames. The individual elements of the object is otherwise the same and resembles that of a typical amplicon data set (otutable, metadata, taxonomy, FASTA sequences, tree). Below is a simple illustration of the data structure and typical workflow of <code>ampvis2</code>:</p>
<p><img src="workflow.png"></p>
</div>
<div class="section level3">
<h3 id="filtering-and-subsetting">Filtering and subsetting<a class="anchor" aria-label="anchor" href="#filtering-and-subsetting"></a>
</h3>
<p>With the <code>ampvis2</code> package comes a large example data set with 573 samples taken from the activated sludge from 55 Danish Wastewater Treatment Plants in the period 2006-2013, which can be loaded with <code>data("MiDAS")</code>. Simply typing the name of any ampvis2 object in the console will show a short summary of the data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"MiDAS"</span><span class="op">)</span></span>
<span><span class="va">MiDAS</span></span></code></pre></div>
<pre><code><span><span class="co">## ampvis2 object with 5 elements. </span></span>
<span><span class="co"><span style="text-decoration: underline;">## Summary of OTU table:</span></span></span>
<span><span class="co"><span style="text-decoration: underline;">## </span>     Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads </span></span>
<span><span class="co">##          658        14969     20890850        10480        46264        31800 </span></span>
<span><span class="co">##    Avg#Reads </span></span>
<span><span class="co">##     31749.01 </span></span>
<span><span class="co">## </span></span>
<span><span class="co"><span style="text-decoration: underline;">## Assigned taxonomy:</span></span></span>
<span><span class="co"><span style="text-decoration: underline;">## </span>      Kingdom        Phylum         Class         Order        Family </span></span>
<span><span class="co">##   14969(100%) 14477(96.71%) 12737(85.09%) 11470(76.63%)  9841(65.74%) </span></span>
<span><span class="co">##         Genus       Species </span></span>
<span><span class="co">##   7380(49.3%)     28(0.19%) </span></span>
<span><span class="co">## </span></span>
<span><span class="co"><span style="text-decoration: underline;">## Metadata variables:</span> 5 </span></span>
<span><span class="co">##  SampleID, Plant, Date, Year, Period</span></span></code></pre>
<p>If you have loaded the raw DNA sequences of the OTUs from a FASTA file (with the <code>fasta =</code> argument) you can also get a short summary by typing the name of the ampvis2 object followed by <code>$refseq</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MiDAS</span><span class="op">$</span><span class="va">refseq</span></span></code></pre></div>
<pre><code><span><span class="co">## 14969 DNA sequences in binary format stored in a list.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Mean sequence length: 472.922 </span></span>
<span><span class="co">##    Shortest sequence: 425 </span></span>
<span><span class="co">##     Longest sequence: 525 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Labels:</span></span>
<span><span class="co">## OTU_1</span></span>
<span><span class="co">## OTU_2</span></span>
<span><span class="co">## OTU_3</span></span>
<span><span class="co">## OTU_4</span></span>
<span><span class="co">## OTU_5</span></span>
<span><span class="co">## OTU_6</span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Base composition:</span></span>
<span><span class="co">##     a     c     g     t </span></span>
<span><span class="co">## 0.261 0.225 0.319 0.194 </span></span>
<span><span class="co">## (Total: 7.08 Mb)</span></span></code></pre>
<p>The loaded data can be subsetted based on variables in the metadata using the <code><a href="../reference/amp_filter_samples.html">amp_subset_samples()</a></code> function, which can then be stored as a new object and analysed separately:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MiDASsubset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_filter_samples.html">amp_subset_samples</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDAS</span>,</span>
<span>  <span class="va">Plant</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Aalborg West"</span>, <span class="st">"Aalborg East"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 590 samples and 5512 OTUs have been filtered </span></span>
<span><span class="co">## Before: 658 samples and 14969 OTUs</span></span>
<span><span class="co">## After: 68 samples and 9457 OTUs</span></span></code></pre>
<p>or for a more complex subset, you can subset based on two or more variables using “&amp;” to separate the conditions, or simply use the function more than once. The “!” (logical NOT operator) can be thought of as “except” and is useful to remove fx outliers. Furthermore, the <code>minreads = 10000</code> argument removes any sample(s) with total amount of reads below the chosen threshold:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MiDASsubset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_filter_samples.html">amp_subset_samples</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDAS</span>,</span>
<span>  <span class="va">Plant</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Aalborg West"</span>, <span class="st">"Aalborg East"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">SampleID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"16SAMP-749"</span><span class="op">)</span>,</span>
<span>  minreads <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 591 samples and 5539 OTUs have been filtered </span></span>
<span><span class="co">## Before: 658 samples and 14969 OTUs</span></span>
<span><span class="co">## After: 67 samples and 9430 OTUs</span></span></code></pre>
<p>The <code><a href="../reference/amp_filter_taxa.html">amp_subset_taxa()</a></code> function instead subsets based on the taxonomy, where you simply provide a vector with the taxa you are interested in, separated by a comma:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MiDAS_Chloroflexi_Actinobacteria</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_filter_taxa.html">amp_subset_taxa</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDAS</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p__Chloroflexi"</span>, <span class="st">"p__Actinobacteria"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 12245 OTUs have been filtered </span></span>
<span><span class="co">## Before: 14969 OTUs</span></span>
<span><span class="co">## After: 2724 OTUs</span></span></code></pre>
<p>The taxonomic rank is indicated by fx "p__" for phylum and "g__" for genus etc, followed by the name of the taxon (case-sensitive, first letter almost always capital). To filter individual OTUs simply provide the OTU name(s) as-is in a vector, fx <code>c("OTU_1206")</code>.</p>
</div>
<div class="section level2">
<h2 id="heatmap">Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h2>
<p>All ampvis2 plots are generated using the <a href="http://ggplot2.tidyverse.org/" class="external-link">ggplot2</a> package. You can change the look of the plots to better suit your needs, add more layers to the plots and use other ggplot2 functions in combination with ampvis plots if needed. Refer to the <a href="http://ggplot2.tidyverse.org/reference/index.html" class="external-link">ggplot2 documentation</a> for more information. <code><a href="../reference/amp_heatmap.html">amp_heatmap()</a></code> by default aggregates to phylum level and shows the top 10 phyla, ordered by mean read abundance across all samples:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/amp_heatmap.html">amp_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDASsubset</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"Plant"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>There are many arguments you can use to suit your needs, for a full list see the reference (click the function names to go to its reference page). For example, you can manually select the level at which to aggregate, how many to show, add additional higher level taxonomic information, group the samples differently by the metadata, hide the values, change the colors and scaling, and much more. You can also adjust the text labels for better readability or adjust the positioning of the legend (adjusting ggplot2 plots is always done with “+” after the actual ampvis2 function):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/amp_heatmap.html">amp_heatmap</a></span><span class="op">(</span><span class="va">MiDASsubset</span>,</span>
<span>            group_by <span class="op">=</span> <span class="st">"Plant"</span>,</span>
<span>            facet_by <span class="op">=</span> <span class="st">"Year"</span>,</span>
<span>            tax_aggregate <span class="op">=</span> <span class="st">"Genus"</span>,</span>
<span>            tax_add <span class="op">=</span> <span class="st">"Phylum"</span>,</span>
<span>            tax_show <span class="op">=</span> <span class="fl">25</span>,</span>
<span>            color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"darkred"</span><span class="op">)</span>,</span>
<span>            plot_colorscale <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>            plot_values <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, size<span class="op">=</span><span class="fl">10</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>        axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,</span>
<span>        legend.position<span class="op">=</span><span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="boxplot">Boxplot<a class="anchor" aria-label="anchor" href="#boxplot"></a>
</h2>
<p><code><a href="../reference/amp_boxplot.html">amp_boxplot()</a></code> generates boxplots, again ordered by mean read abundance across all samples:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/amp_boxplot.html">amp_boxplot</a></span><span class="op">(</span><span class="va">MiDASsubset</span><span class="op">)</span></span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>The arguments you can provide are similar to those used in <code><a href="../reference/amp_heatmap.html">amp_heatmap()</a></code> and other ampvis2 functions:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/amp_boxplot.html">amp_boxplot</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDASsubset</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"Period"</span>,</span>
<span>  tax_show <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  tax_add <span class="op">=</span> <span class="st">"Phylum"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="ordination">Ordination<a class="anchor" aria-label="anchor" href="#ordination"></a>
</h2>
<p>The <code><a href="../reference/amp_ordinate.html">amp_ordinate()</a></code> function has been expanded to support 7 different ordination methods, various data transformations and interactive plots by using <a href="https://plot.ly/" class="external-link">Plotly</a>. By default any OTU with an abundance no higher than 0.1% in any sample is removed, which drastically improves the calculation time. You can of course adjust this threshold manually by changing the <code>filter_species = 0.1</code> argument. Other than this, there are only four main arguments that are involved in the actual calculations, the rest are just various plotting features. These four are <code>type = ""</code>, <code>transform = ""</code>, <code>distmeasure = ""</code>, and lastly <code>constrain = ""</code> for constrained ordination (only used in Redundancy Analysis (RDA) or Canonical Correspondence Analysis (CCA)).</p>
<p>When analysing microbial community composition data it is recommended to use the Hellinger transformation (see <a href="https://doi.org/10.1007/s004420100716" class="external-link">Legendre et al, 2001</a> or <a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0" class="external-link">Numerical Ecology</a> for details) for most types of ordination methods except the distance-based ordination methods (Principal Coordinates Analysis (PCoA) and non-Metric Multidimensional Scaling (nMDS)), where you also have to select a distance measure manually by the <code>distmeasure = ""</code> argument, for example Bray-Curtis dissimilarities:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDASsubset</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"pcoa"</span>,</span>
<span>  distmeasure <span class="op">=</span> <span class="st">"bray"</span>,</span>
<span>  sample_color_by <span class="op">=</span> <span class="st">"Plant"</span>,</span>
<span>  sample_colorframe <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sample_colorframe_label <span class="op">=</span> <span class="st">"Plant"</span></span>
<span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"blank"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>It is also possible to track changes over time by setting the <code>sample_trajectory</code> and <code>sample_trajectory_group</code> arguments, to for example reveal temporal patterns per WWTP:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDASsubset</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"pcoa"</span>,</span>
<span>  distmeasure <span class="op">=</span> <span class="st">"bray"</span>,</span>
<span>  sample_color_by <span class="op">=</span> <span class="st">"Plant"</span>,</span>
<span>  sample_colorframe_label <span class="op">=</span> <span class="st">"Plant"</span>, </span>
<span>  sample_trajectory <span class="op">=</span> <span class="st">"Date"</span>, </span>
<span>  sample_trajectory_group <span class="op">=</span> <span class="st">"Plant"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>And lastly an example of constrained ordination by Canonical Correspondence Analysis (CCA), which in this case reveals how samples taken at different seasonal periods (<code>constrain = "Period"</code>) of the year can explain the data:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ordinationresult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amp_ordinate.html">amp_ordinate</a></span><span class="op">(</span></span>
<span>  <span class="va">MiDASsubset</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"CCA"</span>,</span>
<span>  constrain <span class="op">=</span> <span class="st">"Period"</span>,</span>
<span>  transform <span class="op">=</span> <span class="st">"Hellinger"</span>,</span>
<span>  sample_color_by <span class="op">=</span> <span class="st">"Period"</span>,</span>
<span>  sample_shape_by <span class="op">=</span> <span class="st">"Plant"</span>,</span>
<span>  sample_colorframe <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sample_colorframe_label <span class="op">=</span> <span class="st">"Period"</span>,</span>
<span>  detailed_output <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ordinationresult</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="ampvis2_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Notice that the plot can be saved as a more detailed object (<code>detailed_output = TRUE</code>), so additional details about the ordination result can be obtained for evaluation, fx a screeplot by <code>ordinationresult$screeplot</code> or some more raw data by <code>ordinationresult$model</code>.</p>
</div>
<div class="section level2">
<h2 id="read-more">Read more<a class="anchor" aria-label="anchor" href="#read-more"></a>
</h2>
<p>There are numerous other functions to try out as well as complete documentation for each function highlighted above. Go explore in the <a href="../reference/index.html">Functions</a> tab.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kasper Skytte Andersen, Mads Albertsen, Rasmus Hansen Kirkegaard.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'b0ba3a3eb751685982ccf010113f380a',
    indexName: 'ampvis2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
